### PR TITLE
Separate --strict-warnings and --debug

### PR DESCRIPTION
**Pull request by ekasper**
Use the former for compile-time errors (with the caveat that it also
enables PEDANTIC codepaths), and the latter for runtime debugging.

Remove individual NDEBUG guards now that NDEBUG is always defined in
release builds.

In Travis, do --strict-warnings on BUILDONLY configurations. This
ensures that the tests run even if --strict-warnings fail, and avoids
hiding unrelated test failures.

Also fix a couple of warnings with clang 3.8 (one of them false positive).

### CODE REVIEW

File: CHANGES [hunk 1]
```diff
@@ -4,6 +4,11 @@
 
  Changes between 1.0.2g and 1.1.0  [xx XXX xxxx]
 
+  *) --strict-warnings no longer enables runtime debugging options
+     such as REF_DEBUG. Instead, REF_DEBUG is enabled automatically
```
Comment (richsalz) [inline CHANGES]: Should "REF_DEBUG is" be "they are"?  Or is it only REF_DEBUG?

File: Configure [hunk 2]
```diff
@@ -76,8 +76,12 @@ my $usage="Usage: Configure [no-<cipher> ...] [enable-<cipher> ...] [-Dxxx] [-lx
 
 # Minimum warning options... any contributions to OpenSSL should at least get
 # past these.
-
-my $gcc_devteam_warn = "-DPEDANTIC -DREF_DEBUG -DDEBUG_UNUSED -DBIO_DEBUG"
+#
+# DEBUG_UNUSED enables __owur (warn unused result) checks.
+# PEDANTIC conditionally enables different codepaths in low-level crypto code
+# to get past PEDANTIC. Contributions to OpenSSL should therefore also pass
+# without PEDANTIC.
```
Comment (richsalz) [inline Configure]: PEDANTIC also needed for some compilers that can't handle files that are "empty"  Not sure if that's worth calling out.

File: test/ct_test.c [hunk 3]
```diff
@@ -101,6 +101,8 @@ static CT_TEST_FIXTURE set_up(const char *const test_case_name)
 {
     CT_TEST_FIXTURE fixture;
     int setup_ok = 1;
+    memset(&fixture, 0, sizeof(fixture));
```
Comment (richsalz) [inline test/ct_test.c]: blank line before the memset, not after :)

File: CHANGES (hunk 1 above)
Comment (ekasper) [inline CHANGES]: REF_DEBUG is the only one moved from --strict-warnings to --debug (BIO_DEBUG was just deleted). But the change is meant to stick for the future as well. Not sure how to word it better.

Comment (richsalz) [inline CHANGES]: Me either. :(  We should turn on *_DEBUG if --debug is built.  But a quick grep shows that the naming of DEBUG ifdef's is not consistent (ENGINE_REF_COUNT_DEBUG, OPENSSL_DEBUG_PKCS5V2, 
OPENSSL_(INIT|POLICY)_DEBUG, etc.  Ugh.  Separate MR and discussion about this.

File: Configure (hunk 2 above)
Comment (ekasper) [inline Configure]: That sounds harmless. The gist of the comment here is that PEDANTIC actually changes runtime behaviour. We should really have tests for both --strict-warnings and --strict-warnings-with-pedantic as two different configurations (because you want to build the non-pedantic codepaths with the other warnings as well).

(I fixed up the comment a little.)

**Comment by richsalz** [comment]
yes PEDANTIC is too broad a paintbrush, but that's a separate issue from this.
+1

File: test/ct_test.c (hunk 3 above)
Comment (ekasper) [inline test/ct_test.c]: Fixed.

File: CHANGES (hunk 1 above)
Comment (ekasper) [inline CHANGES]: Indeed, there's more to follow up with here. Thanks for the review; levitte@ wanted to take a look as well so I'll wait, in case he has good ideas about cleaning up --debug. Right now what you get seems to depend on the compiler. (The various confs in 10-main.conf have different debug pickers.)

**Comment by ekasper** [comment]
Rebased on top of Andy's changes. Please re-review.

File: apps/s_cb.c [hunk 4]
```diff
@@ -507,12 +507,12 @@ long bio_dump_callback(BIO *bio, int cmd, const char *argp,
...
     if (cmd == (BIO_CB_READ | BIO_CB_RETURN)) {
         BIO_printf(out, "read from %p [%p] (%lu bytes => %ld (0x%lX))\n",
-                   (void *)bio, argp, (unsigned long)argi, ret, ret);
+                   (void *)bio, (void *)argp, (unsigned long)argi, ret, ret);
         BIO_dump(out, argp, (int)ret);
         return (ret);
     } else if (cmd == (BIO_CB_WRITE | BIO_CB_RETURN)) {
         BIO_printf(out, "write to %p [%p] (%lu bytes => %ld (0x%lX))\n",
-                   (void *)bio, argp, (unsigned long)argi, ret, ret);
+                   (void *)bio, (void *)argp, (unsigned long)argi, ret, ret);
```
Comment (levitte) [inline apps/s_cb.c]: Not sure why this is part of this PR, seems out of scope.

File: test/ct_test.c [hunk 5]
```diff
@@ -101,6 +101,9 @@ static CT_TEST_FIXTURE set_up(const char *const test_case_name)
 {
     CT_TEST_FIXTURE fixture;
     int setup_ok = 1;
+
+    memset(&fixture, 0, sizeof(fixture));
+
```
Comment (levitte) [inline test/ct_test.c]: Dito

File: apps/s_cb.c (hunk 4 above)
Comment (ekasper) [inline apps/s_cb.c]: So I happened to be building with clang-3.8 and they were needed to get --strict-warnings to build. Commit description says, "Also fix a couple of warnings with clang 3.8 (one of them false positive).".

I can separate them into a different commit if it's important.

File: Configure [hunk 6]
```diff
@@ -78,6 +78,7 @@ my $usage="Usage: Configure [no-<cipher> ...] [enable-<cipher> ...] [-Dxxx] [-lx
 # past these.
 
 # DEBUG_UNUSED enables __owur (warn unused result) checks.
+
```
Comment (ekasper) [inline Configure]: This is a merge fail and I'll fix before pushing.

File: apps/s_cb.c (hunk 4 above)
Comment (levitte) [inline apps/s_cb.c]: Separate commit is cleaner

Comment (ekasper) [inline apps/s_cb.c]: Done.

**Comment by levitte** [comment]
+1 from eyeballing, pending Travis success (which doesn't mean green, but means every job should at least start building object files)

**Comment by ekasper** [comment]
Travis failed, obv. Hang on...

**Comment by levitte** [comment]
> Travis failed

Not in a way that afffects this PR.  Ship it!

**Comment by ekasper** [comment]
No, it _is_ my fault. One's a mixed declarations warning that clang oddly missed but gcc doesn't; and the other is something I knew but forgot: we can't ubsan non-pedantic builds. Fixes coming up.

**Comment by ekasper** [comment]
There. Let's see what Travis thinks of that.

**Comment by ekasper** [comment]
Well, it still fails but now I have no idea why :( https://travis-ci.org/openssl/openssl/jobs/121755663

**Comment by ekasper** [comment]
Note it's completely happy until the very end when it says "Done. Your build exited with 1."

**Comment by levitte** [comment]
That looks like a Travis hickup.  I've restarted the job to make sure

**Comment by levitte** [comment]
Hmmm...  well, no idea what's up with that...

**Comment by ekasper** [comment]
Nah, there's actually another error; it's in the middle of the build script somehow. gcc doesn't like some variables that are only used for an assert. Another attempt coming up.

**Comment by ekasper** [comment]
Green now!

**Comment by levitte** [comment]
+1 reconfirmed
